ST4: 4065+ HTMLSheet Updates

TerminalFi opened the pull request:
**Should be a ST4 only branch**
- Requires => `4074`

Will update the Package Control Channel to support ST3 and ST4 tags and set appropriate ST4 version requirement

### Major Changes

- [x] Removes `2markdown.py` in favor of Mdpopups
  - Provides proper code highlighting and handles md2html
- [x] Removes `markdown2html.py` implements `ImageParser.py`
  - Utilizes 95% of the same code with minor tweaks
- [x] Adds `add_on_change` for monitoring the `SETTING_DELAY_BETWEEN_UPDATES`
to update the delay in real time. (**Could be back ported**)
- [x] Adds `MdlpEraseCommand` for erasing a view when called from `on_pre_close` **_Should probably be implemented on the ST3 branch to fix a bug in which EventListener doesn't have a `edit` object_**
https://github.com/math2001/MarkdownLivePreview/blob/ec27d980a3ebd4c73c23f9036e5b4edf1dea2e27/MarkdownLivePreview.py#L130-L135
- [x] Replaces Phantoms with HTMLSheets which are implemented in ST4065
- [x] Implements a `PREVIEW_VIEWS` hash map to allow for multiple markdown previews (**Could be back ported**)
- [x] Adds a `syntax` setting that allows you to enable the plugin on different syntax files (**Could be back ported**)
- [ ] #22

### Fixes

fixes #94 - Fixed with ST4 Changes. Likely can fix in ST3 by using Mdpopups but not HTMLSheet
fixes #98 via facelessuser/sublime-markdown-popups@5c65d6f
fixes #103
fixes #104
fixes #106

TerminalFi committed:
implement: sublime HTMLSheet
implement: image to base64

TerminalFi committed:
implement: ImageParser.py

TerminalFi committed:
remove: 2markdown.py

TerminalFi committed:
implement: invalid_image

TerminalFi committed:
fix: on_pre_close

jrappen commented:
@TheSecEng: please add to the above message

---

```text
fixes #98 via facelessuser/sublime-markdown-popups@5c65d6f
fixes #103
fixes #104
```

TerminalFi committed:
fix: strip html comments

TheSecEng commented:
Investigating an issue with `mdpopups` where it doesn't convert some Markdown correctly

TerminalFi committed:
fix: passing html and stating it was MD

TerminalFi committed:
implement: render checkbox with unicode

TerminalFi committed:
fix: BS4 kept adding "closing <br>"
fix: strip comments
fix: frontmatter because mdpopups wont support nl2br soon
fix: trying to update non-existent previews

TerminalFi committed:
implement: multiple syntax support via `syntax` setting

jrappen commented on MarkdownLivePreview.py:
```diff
@@ -78,11 +141,8 @@ def run(self, edit):
-        preview_view.set_name("Preview")
-        # FIXME: hide number lines on preview
+
+        preview_view = mdpopups.new_html_sheet(preview_window, "Preview", "")
```
I suggest changing this to:

`preview_view = mdpopups.new_html_sheet(window=preview_window, name="Preview", contents="")`

jrappen commented on MarkdownLivePreview.py:
```diff
@@ -169,72 +218,76 @@ def on_close(self, markdown_view):
+            html_content, basepath, partial(self._update_preview, view), resources,
         )
 
+        mdpopups.update_html_sheet(preview_view, html_content, md=False)
```
I suggest changing this to:

`mdpopups.update_html_sheet(sheet=preview_view, contents=html_content, md=False)`

jrappen commented:
ST Build 4074 and mdpopups `v3.7.0` had breaking changes due to Sublime no longer supporting `cmd` and `args` for `sublime.Window.new_html_sheet`, at least breaking the `4065-4073` behavior.

Code is fine as is, but requirement should be bumped.

TerminalFi committed:
implement: MarkdownLivePreviewBaseCommand for shared functions
implement: positional argument names

TerminalFi committed:
fix: delay logic

TerminalFi committed:
format: ImageParser

TerminalFi committed:
remove: unneeded markdown frontmatter

TerminalFi committed:
implement: render_checkbox font size

rwols commented on ImageParser.py:
```diff
@@ -0,0 +1,135 @@
+# FIXME: how do I choose how many workers I want?
+# - Does thread pool reuse threads or does it stupidly throw them out?
+# - (we could implement something of our own)
+executor = concurrent.futures.ThreadPoolExecutor(max_workers=5)
```
This is overkill for this sort of plugin.

rwols commented on ImageParser.py:
```diff
@@ -0,0 +1,135 @@
+            )
+        return "{}{}".format(
+            "data:image/png;base64,", b64encode(image_content).decode("utf-8")
+        )
```
Looks like you want to use asyncio instead of a ThreadPoolExecutor (which is IMO completely the wrong tool for this job).

AmjadHD commented:
Why did work on this stopped ?

TerminalFi commented:
> Why did work on this stopped ?



> Why did work on this stopped ?

I just am in the middle of moving and what not. I’d be happy to get help with it. 

I’ll start working on it again when I have time.
